File: config/src/main/scala/com/typesafe/config/impl/SubstitutionExpression.scala
package com.typesafe.config.impl

final class SubstitutionExpression(val path: Path, val optional: Boolean) {

    private[impl] def changePath(newPath: Path) =
        if (newPath eq path) this else new SubstitutionExpression(newPath, optional)

    override def toString: String =
        "${" + (if (optional) "?" else "") + path.render + "}"

    override def equals(other: Any): Boolean =
        if (other.isInstanceOf[SubstitutionExpression]) {
            val otherExp =
                other.asInstanceOf[SubstitutionExpression]
            otherExp.path == this.path && otherExp.optional == this.optional
        } else false

    override def hashCode: Int = {
        var h = 41 * (41 + path.hashCode)
        h = 41 * (h + (if (optional) 1 else 0))
        h
    }
}

File: config/src/main/java/com/typesafe/config/impl/SubstitutionExpression.java
package com.typesafe.config.impl;

final class SubstitutionExpression {

    final private Path path;
    final private boolean optional;
    final private boolean listExpansion;

    SubstitutionExpression(Path path, boolean optional) {
        this(path, optional, false);
    }

    SubstitutionExpression(Path path, boolean optional, boolean listExpansion) {
        this.path = path;
        this.optional = optional;
        this.listExpansion = listExpansion;
    }

    Path path() {
        return path;
    }

    boolean optional() {
        return optional;
    }

    boolean listExpansion() {
        return listExpansion;
    }

    SubstitutionExpression changePath(Path newPath) {
        if (newPath == path)
            return this;
        else
            return new SubstitutionExpression(newPath, optional, listExpansion);
    }

    SubstitutionExpression changeListExpansion(boolean newListExpansion) {
        if (newListExpansion == listExpansion)
            return this;
        else
            return new SubstitutionExpression(path, optional, newListExpansion);
    }

    @Override
    public String toString() {
        return "${" + (optional ? "?" : "") + path.render() + (listExpansion ? "[]" : "") + "}";
    }

    @Override
    public boolean equals(Object other) {
        if (other instanceof SubstitutionExpression) {
            SubstitutionExpression otherExp = (SubstitutionExpression) other;
            return otherExp.path.equals(this.path) && otherExp.optional == this.optional
                    && otherExp.listExpansion == this.listExpansion;
        } else {
            return false;
        }
    }

    @Override
    public int hashCode() {
        int h = 41 * (41 + path.hashCode());
        h = 41 * (h + (optional ? 1 : 0));
        h = 41 * (h + (listExpansion ? 1 : 0));
        return h;
    }
}
